File: app/src/main/java/me/tagavari/airmessage/messaging/viewholder/VHMessageComponentLocation.java
package me.tagavari.airmessage.messaging.viewholder;

import android.content.Context;
import android.content.res.ColorStateList;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.material.card.MaterialCardView;
import com.google.android.material.shape.MaterialShapeDrawable;
import com.google.android.material.shape.ShapeAppearanceModel;

import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.subjects.SingleSubject;
import me.tagavari.airmessage.R;
import me.tagavari.airmessage.enums.MessageComponentType;
import me.tagavari.airmessage.helper.MessageShapeHelper;
import me.tagavari.airmessage.view.RoundedFrameLayout;

public class VHMessageComponentLocation extends VHMessageComponentAttachment {
	public final MaterialCardView groupContent;
	public final View viewBorder;
	public final RoundedFrameLayout mapContainer;
	public final MapView mapView;
	public final TextView labelTitle;
	public final TextView labelAddress;
	
	private final SingleSubject<GoogleMap> mapSubject;
	
	public VHMessageComponentLocation(@NonNull View itemView, ViewGroup groupContainer, ViewGroup stickerContainer, ViewGroup tapbackContainer, ViewGroup groupPrompt, TextView labelPromptSize, TextView labelPromptType, ImageView iconPrompt, ViewGroup groupProgress, ProgressBar progressProgress, ImageView iconProgress, ViewGroup groupOpen, TextView labelOpen, ViewGroup groupContentFrame, MaterialCardView groupContent, View viewBorder, RoundedFrameLayout mapContainer, MapView mapView, TextView labelTitle, TextView labelAddress) {
		super(itemView, groupContainer, stickerContainer, tapbackContainer, groupPrompt, labelPromptSize, labelPromptType, iconPrompt, groupProgress, progressProgress, iconProgress, groupOpen, labelOpen, groupContentFrame);
		this.groupContent = groupContent;
		this.viewBorder = viewBorder;
		this.mapContainer = mapContainer;
		this.mapView = mapView;
		this.labelTitle = labelTitle;
		this.labelAddress = labelAddress;
		
		//Initializing the map
		mapView.onCreate(null);
		mapSubject = SingleSubject.create();
		mapView.getMapAsync(mapSubject::onSuccess);
	}
	
	public Single<GoogleMap> getGoogleMap() {
		return mapSubject;
	}
	
	@Override
	void updateContentViewEdges(Context context, boolean anchoredTop, boolean anchoredBottom, boolean alignToRight) {
		//Creating the message shape
		ShapeAppearanceModel messageAppearance = MessageShapeHelper.createRoundedMessageAppearance(context.getResources(), anchoredTop, anchoredBottom, alignToRight);
		
		//Assigning the border shape
		MaterialShapeDrawable borderDrawable = new MaterialShapeDrawable(messageAppearance);
		borderDrawable.setFillColor(ColorStateList.valueOf(0));
		borderDrawable.setStroke(2, context.getColor(R.color.colorDivider));
		viewBorder.setBackground(borderDrawable);
		
		//Assigning the map shape
		mapContainer.setRadiiRaw(MessageShapeHelper.createStandardRadiusArrayTop(context.getResources(), anchoredTop, alignToRight));
		
		//Assigning the card shape
		groupContent.setShapeAppearanceModel(messageAppearance);
	}
	
	@Override
	void updateContentViewColoring(Context context, int colorTextPrimary, int colorTextSecondary, int colorBackground) {
	
	}
	
	@Override
	public int getComponentType() {
		return MessageComponentType.attachmentLocation;
	}
}

File: app/src/main/java/me/tagavari/airmessage/messaging/viewholder/VHMessageComponentLocation.kt
package me.tagavari.airmessage.messaging.viewholder

import android.content.Context
import android.content.res.ColorStateList
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import com.google.android.material.card.MaterialCardView
import com.google.android.material.shape.MaterialShapeDrawable
import io.reactivex.rxjava3.core.Completable
import me.tagavari.airmessage.R
import me.tagavari.airmessage.enums.MessageComponentType
import me.tagavari.airmessage.flavor.VHMessageComponentLocationMap
import me.tagavari.airmessage.helper.MessageShapeHelper.createRoundedMessageAppearance
import me.tagavari.airmessage.helper.MessageShapeHelper.createStandardRadiusArrayTop
import me.tagavari.airmessage.util.LatLngInfo
import me.tagavari.airmessage.view.RoundedFrameLayout

class VHMessageComponentLocation(
	itemView: View,
	groupContainer: ViewGroup,
	stickerContainer: ViewGroup,
	tapbackContainer: ViewGroup,
	groupPrompt: ViewGroup,
	labelPromptSize: TextView,
	labelPromptType: TextView,
	iconPrompt: ImageView,
	groupProgress: ViewGroup,
	progressProgress: ProgressBar,
	iconProgress: ImageView,
	groupOpen: ViewGroup,
	labelOpen: TextView,
	groupContentFrame: ViewGroup,
	val groupContent: MaterialCardView,
	val viewBorder: View,
	val mapContainer: RoundedFrameLayout,
	val mapView: View?, //GMS MapView (if available)
	val labelTitle: TextView,
	val labelAddress: TextView
) : VHMessageComponentAttachment(
	itemView,
	groupContainer,
	stickerContainer,
	tapbackContainer,
	groupPrompt,
	labelPromptSize,
	labelPromptType,
	iconPrompt,
	groupProgress,
	progressProgress,
	iconProgress,
	groupOpen,
	labelOpen,
	groupContentFrame
) {
	override val componentType = MessageComponentType.attachmentLocation
	private val flavorMap = VHMessageComponentLocationMap(this)
	val mapLoadCompletable get() = flavorMap.mapLoadCompletable
	
	override fun updateContentViewEdges(context: Context, anchoredTop: Boolean, anchoredBottom: Boolean, alignToRight: Boolean) {
		//Creating the message shape
		val messageAppearance =
			createRoundedMessageAppearance(context.resources, anchoredTop, anchoredBottom, alignToRight)
		
		//Assigning the border shape
		val borderDrawable = MaterialShapeDrawable(messageAppearance)
		borderDrawable.fillColor = ColorStateList.valueOf(0)
		borderDrawable.setStroke(2f, context.getColor(R.color.colorDivider))
		viewBorder.background = borderDrawable
		
		//Assigning the map shape
		mapContainer.setRadiiRaw(createStandardRadiusArrayTop(context.resources, anchoredTop, alignToRight))
		
		//Assigning the card shape
		groupContent.shapeAppearanceModel = messageAppearance
	}
	
	override fun updateContentViewColoring(context: Context, colorTextPrimary: Int, colorTextSecondary: Int, colorBackground: Int) = Unit
	
	fun setMapLocation(location: LatLngInfo?) {
		flavorMap.setMapLocation(itemView.context, location)
	}
}